src/FSharp.Core/immutablearray.fs

```diff
@@ -0,0 +1,490 @@
+// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.FSharp.Collections
+
+open System
+open System.Collections.Generic
+open System.Collections.Immutable
+
+open Microsoft.FSharp.Core
+open Microsoft.FSharp.Core.CompilerServices
+open Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicOperators
+
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+[<RequireQualifiedAccess>]
+module ImmutableArray =
+
+    [<CompiledName("Length")>]
+    let inline length (array: ImmutableArray<'T>) : int =
+        array.Length
+
+    [<CompiledName("Singleton")>]
+    let inline singleton (item: 'T) : ImmutableArray<_> =
+        ImmutableArray.Create(item)
+
+    [<CompiledName("Initialize")>]
+    let inline init n ([<InlineIfLambda>] f: int -> 'T) : ImmutableArray<_> =
+        match n with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> ImmutableArray.Create(f 0)
+        | n ->
+            if n < 0 then
+                invalidArg "n" "Below zero."
+
+            let builder = ImmutableArray.CreateBuilder(n)
+            for i = 0 to n - 1 do
+                builder.Add(f i)
+            builder.MoveToImmutable()
+
+    [<CompiledName("ZeroCreate")>]
+    let inline zeroCreate (count: int) : ImmutableArray<'T> =
+        if count < 0 then
+            invalidArgInputMustBeNonNegative (nameof(count)) count
+
+        init count (fun _ -> LanguagePrimitives.GenericZero)
+
+    [<CompiledName("Create")>]
+    let inline create (count: int) (value: 'T) : ImmutableArray<'T> =
+        if count < 0 then
+            invalidArgInputMustBeNonNegative (nameof(count)) count
+
+        init count (fun _ -> value)
+
+    [<CompiledName("IsEmpty")>]
+    let inline isEmpty (arr: ImmutableArray<_>) = arr.IsEmpty
+
+    [<CompiledName("Empty")>]
+    [<GeneralizableValue>]
+    let empty<'T> = ImmutableArray<'T>.Empty
+
+    let inline concatImmutableArrays (arrs: ImmutableArray<ImmutableArray<'T>>) : ImmutableArray<'T> =
+        match arrs.Length with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> arrs[0]
+        | 2 -> arrs[0].AddRange(arrs[1])
+        | n ->
+            let mutable acc = 0
+
+            for h in arrs do
+                acc <- acc + h.Length
+
+            let builder = ImmutableArray.CreateBuilder(acc)
+
+            for i = 0 to n - 1 do
+                builder.AddRange(arrs[i])
+
+            builder.MoveToImmutable()
+
+    [<CompiledName("Concat")>]
+    let inline concat (arrs: IEnumerable<ImmutableArray<'T>>) : ImmutableArray<'T> =
+        match arrs with
+        | :? ImmutableArray<ImmutableArray<'T>> as arrs -> concatImmutableArrays arrs
+        | arrs -> concatImmutableArrays (ImmutableArray.CreateRange(arrs))
```

Comment:
If you include the body of concatImmutableArrays in this function then collect can call concat.  Which eliminates a wierd public api: concatImmutableArrays